pkg/wasm/cache.go

```diff
@@ -104,25 +106,41 @@ func NewLocalFileCache(dir string, purgeInterval, moduleExpiry time.Duration, in
 	return cache
 }
 
+func needToUseCache(pullPolicy extensions.PullPolicy, u *url.URL) bool {
+	if u.Scheme == "oci" {
+		switch pullPolicy {
+		case extensions.PullPolicy_Always:
+			return false
+		case extensions.PullPolicy_IfNotPresent:
+			return true
+		default:
+			return !strings.HasSuffix(u.Path, ":latest")
+		}
+	}
+	return true
+}
+
 // Get returns path the local Wasm module file.
-func (c *LocalFileCache) Get(downloadURL, checksum string, timeout time.Duration, pullSecret []byte) (string, error) {
+func (c *LocalFileCache) Get(downloadURL, checksum string, timeout time.Duration, pullSecret []byte, pullPolicy extensions.PullPolicy) (string, error) {
 	// Construct Wasm cache key with downloading URL and provided checksum of the module.
 	key := cacheKey{
 		downloadURL: downloadURL,
 		checksum:    checksum,
 	}
 
-	// First check if the cache entry is already downloaded.
-	if modulePath := c.getEntry(key); modulePath != "" {
-		return modulePath, nil
-	}
-
-	// If not, fetch images.
 	u, err := url.Parse(downloadURL)
 	if err != nil {
 		return "", fmt.Errorf("fail to parse Wasm module fetch url: %s", downloadURL)
 	}
 
+	// First check if the cache entry is already downloaded and policy does not require to pull always.
+	if needToUseCache(pullPolicy, u) {
+		if modulePath := c.getEntry(key); modulePath != "" {
+			return modulePath, nil
+		}
+	}
+	// If not, fetch images.
```

Comment:
I think Kubernetes logic is a bit more efficient here. Even on Always it will not always re-pull the image and will instead just do a check against the current tag and make sure the sha256 of it matches what we have locally. So in the case it hasn't changed, it will be pretty cheap - just pulling the config of the image, not the full image.

Reply:
Good point. will take the way.

Reply:
This is resolved by https://github.com/istio/istio/pull/38630